src/iamserver/nodecontroller.cpp

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <thread>
+
+#include "log.hpp"
+#include "nodecontroller.hpp"
+#include "utils/convert.hpp"
+
+/***********************************************************************************************************************
+ * NodeStreamHandler
+ **********************************************************************************************************************/
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+NodeStreamHandler::NodeStreamHandler(const std::vector<aos::NodeStatus>& allowedStatuses,
+    NodeServerReaderWriter* stream, grpc::ServerContext* context, aos::iam::nodemanager::NodeManagerItf* nodeManager)
+    : mAllowedStatuses(allowedStatuses)
+    , mStream(stream)
+    , mContext(context)
+    , mNodeManager(nodeManager)
+{
+}
+
+NodeStreamHandler::~NodeStreamHandler()
+{
+    Close();
+}
+
+void NodeStreamHandler::Close()
+{
+    if (mIsClosed.exchange(true)) {
+        return;
+    }
+
+    std::lock_guard lock {mMutex};
+
+    mContext->TryCancel();
+
+    mPendingMessages.clear();
+}
+
+std::string NodeStreamHandler::GetNodeID() const
+{
+    std::lock_guard lock {mNodeIDMutex};
+
+    return mNodeID;
+}
+
+aos::Error NodeStreamHandler::HandleStream()
+{
+    LOG_DBG() << "Process stream handler";
+
+    aos::Error                    err = aos::ErrorEnum::eNone;
+    iamproto::IAMOutgoingMessages outgoing;
+
+    while (mStream->Read(&outgoing)) {
+        LOG_DBG() << "Receive message: type=" << outgoing.GetTypeName().c_str();
+
+        const auto messageCase = outgoing.IAMOutgoingMessage_case();
+        if (messageCase == iamproto::IAMOutgoingMessages::IAMOutgoingMessageCase::IAMOUTGOINGMESSAGE_NOT_SET) {
+            continue;
+        }
+
+        if (outgoing.has_node_info()) {
+            if (auto err = HandleNodeInfo(outgoing.node_info()); !err.IsNone()) {
```

Comment:
You shouldn't break the loop in this case I guess. It is internal error not communication error.

Reply:
Error is possible if secondary node is provisioned but connected to public IAM main (or uprovisioned to protected)
Also error is generated if protobuf conversion to AOS fails
Or node manager fails

Looks like we should notify secondary IAM somehow that error occurred.